cloudwatchlogs/cloudwatchlogs_lambda.js

```diff
@@ -0,0 +1,215 @@
+    
+    if (sourceCategoryOverride !== null && sourceCategoryOverride !== '' && sourceCategoryOverride != 'none') {
+        sourceCategory = sourceCategoryOverride;
+    }
+    
+    if (sourceHostOverride !== null && sourceHostOverride !== '' && sourceHostOverride != 'none') {
+        sourceHost = sourceHostOverride;
+    } else {
+        sourceHost = awslogsData.logGroup;
+    }
+    
+    if (sourceNameOverride !== null && sourceNameOverride !== '' && sourceNameOverride != 'none') {
+        sourceName = sourceNameOverride;
+    } else {
+        sourceName = awslogsData.logStream;
+    }
+    
+    // Ability to override metadata within the message
+    // Useful within Lambda function console.log to dynamically set metadata fields within SumoLogic.
+    if (message.hasOwnProperty('_sumo_metadata')) {
+        var metadataOverride = message._sumo_metadata;
+        if (metadataOverride.category) {
+            sourceCategory = metadataOverride.category;
+        }
+        if (metadataOverride.host) {
+            sourceHost = metadataOverride.host;
+        }
+        if (metadataOverride.source) {
+            sourceName = metadataOverride.source;
+        }
+        delete message._sumo_metadata;
+    }
+    return sourceName + ':' + sourceCategory + ':' + sourceHost;
+    
+}
+
+function postToSumo(context, messages) {
+    var messagesTotal = Object.keys(messages).length;
+    var messagesSent = 0;
+    var messageErrors = [];
+    
+    var urlObject = url.parse(SumoURL);
+    var options = {
+        'hostname': urlObject.hostname,
+        'path': urlObject.pathname,
+        'method': 'POST'
+    };
+    
+    var finalizeContext = function () {
+        var total = messagesSent + messageErrors.length;
+        if (total == messagesTotal) {
+            console.log('messagesSent: ' + messagesSent + ' messagesErrors: ' + messageErrors.length);
+            if (messageErrors.length > 0) {
+                context.fail('errors: ' + messageErrors);
+            } else {
+                context.succeed();
+            }
+        }
+    };
+    
+    
+    Object.keys(messages).forEach(function (key, index) {
+        var headerArray = key.split(':');
+        
+        options.headers = {
+            'X-Sumo-Name': headerArray[0],
+            'X-Sumo-Category': headerArray[1],
+            'X-Sumo-Host': headerArray[2]
+        };
+        
+        var req = https.request(options, function (res) {
+            var body = '';
+            res.setEncoding('utf8');
+            res.on('data', function (chunk) {
+                body += chunk;
+            });
+            res.on('end', function () {
+                if (res.statusCode == 200) {
+                    messagesSent++;
+                } else {
+                    errors.push('HTTP Return code ' + res.statusCode);
+                }
+                finalizeContext();
+            });
+        });
+        
+        req.on('error', function (e) {
```

Comment:
my understanding is that there's definitely times in which we return an error - (in)famously when throttling (429). it would be, like, awesome if we could react to this and somehow find a way to re-issue the request later. I don't know if this is something that Lambda knows how to do? maybe we need an SQS queue and another Lambda for that?